### PR TITLE
Incorporate new Spark DataFrame API

### PR DESCRIPTION
**Pull request by cpcloud**

### CODE REVIEW

File: odo/backends/sparksql.py [hunk 1]
```diff
@@ -58,19 +91,55 @@ def rdd_to_sqlcontext(ctx, rdd, name=None, dshape=None, **kwargs):
         dshape = dshape.measure
     sql_schema = dshape_to_schema(dshape)
     sdf = ctx.applySchema(rdd, sql_schema)
-    ctx.registerRDDAsTable(sdf, name or next(_names))
+    if name is None:
+        name = next(_names)
+    ctx.registerDataFrameAsTable(sdf, name)
+    ctx.cacheTable(name)
```
Comment (cpcloud) [inline odo/backends/sparksql.py]: I've decided to cache by default. This enables Spark to do what we call "lean projection" in blaze.

**Comment by cpcloud** [comment]
@mrocklin can you review pls when you get a chance? thanks!

File: .travis.yml [hunk 2]
```diff
@@ -37,7 +37,7 @@ install:
   - pip install --upgrade git+git://github.com/cpcloud/redshift_sqlalchemy
 
   # install pyspark
-  - if [[ $TRAVIS_PYTHON_VERSION == '2.7' ]]; then conda install -c blaze spark; fi
+  - if [[ $TRAVIS_PYTHON_VERSION == '2.7' ]]; then conda install -c blaze -c https://conda.binstar.org/blaze/channel/dev spark; fi
```
Comment (mrocklin) [inline .travis.yml]: What is in dev channel that isn't in main?  spark 1.3?

File: odo/backends/sparksql.py [hunk 3]
```diff
@@ -42,10 +56,29 @@ def jsonlines_to_sparksql(ctx, json, dshape=None, name=None, schema=None,
...
+
+
+@convert.register(base, SparkDataFrame, cost=200.0)
+def spark_df_to_base(df, **kwargs):
+    return df.collect()[0][0]
+
+
+@convert.register(pd.DataFrame, SparkDataFrame, cost=200.0)
+def spark_df_to_pandas_df(df, **kwargs):
+    return df.toPandas()
```
Comment (mrocklin) [inline odo/backends/sparksql.py]: Is this faster than convert to list above?  That choice might be relevant.

Comment (mrocklin) [inline odo/backends/sparksql.py]: Even if it isn't faster it's probably better to give it a lower cost if this edge is more robust.

File: odo/backends/sparksql.py [hunk 4]
```diff
@@ -58,19 +91,55 @@ def rdd_to_sqlcontext(ctx, rdd, name=None, dshape=None, **kwargs):
...
+
+def chunk_file(filename, chunksize=1 << 30):
+    """Stream `filename` in chunks of size `chunksize`.
+
+    Parameters
+    ----------
+    filename : str
+        File to chunk
+    chunksize : int, optional, defaults to 1GB
+        Number of bytes to hold in memory at a single time
```
Comment (mrocklin) [inline odo/backends/sparksql.py]: Suggest smaller chunks.  My laptop has four cores and four gigabytes of ram.  Unless there is a significant performance difference I'd suggest dropping an order of magnitude.  If there is no performance difference then maybe drop a couple.

File: odo/backends/sparksql.py [hunk 5]
```diff
@@ -58,19 +91,55 @@ def rdd_to_sqlcontext(ctx, rdd, name=None, dshape=None, **kwargs):
...
+    with open(filename, mode='rb') as f:
+        for chunk in iter(partial(f.read, chunksize), ''):
+            yield chunk
+
+
+@append.register((JSONLines, JSON), SparkDataFrame)
+def spark_df_to_jsonlines(js, df, pattern='part-*', **kwargs):
+    tmpd = tempfile.mkdtemp()
+    try:
+        df.save(tmpd, source='org.apache.spark.sql.json', mode='overwrite')
```
Comment (mrocklin) [inline odo/backends/sparksql.py]: What are the options for source here?

In what format does this write JSON, line delimited or traditional?

File: .travis.yml (hunk 2 above)
Comment (cpcloud) [inline .travis.yml]: Yep

Comment (mrocklin) [inline .travis.yml]: Should that be in main cc @quasiben

File: odo/backends/sparksql.py (hunk 5 above)
Comment (cpcloud) [inline odo/backends/sparksql.py]: They're taken from here: https://github.com/apache/spark/tree/master/sql/core/src/main/scala/org/apache/spark/sql

The ones that work are `json`, `jdbc`, and `parquet`

Comment (cpcloud) [inline odo/backends/sparksql.py]: It writes JSON as a directory of newline delimited JSON, one for each partition.

More info here: http://stackoverflow.com/a/23902886/564538

Comment (mrocklin) [inline odo/backends/sparksql.py]: Should probably restrict this to just `JSONLines` then

Comment (cpcloud) [inline odo/backends/sparksql.py]: ah right, my mistake

File: odo/backends/sparksql.py (hunk 3 above)
Comment (cpcloud) [inline odo/backends/sparksql.py]: It's a little bit faster. Lowering the cost now.

File: .travis.yml (hunk 2 above)
Comment (quasiben) [inline .travis.yml]: For now in Dev while I do some house keeping in conda-cluster

Comment (quasiben) [inline .travis.yml]: For now in dev while I do some house keeping in conda-cluster

File: odo/backends/sparksql.py (hunk 4 above)
Comment (cpcloud) [inline odo/backends/sparksql.py]: Good idea. I shaved off about 45 seconds in some runs by lowering the chunk size:

```python
In [5]: time foo = odo(df, 'jsonlines://foo.json', chunksize=1 << 30)
CPU times: user 186 ms, sys: 4.47 s, total: 4.65 s
Wall time: 2min

In [6]: time foo = odo(df, 'jsonlines://foo.json', chunksize=1 << 27)
CPU times: user 147 ms, sys: 4.76 s, total: 4.9 s
Wall time: 1min 13s

In [7]: time foo = odo(df, 'jsonlines://foo.json', chunksize=1 << 25)
CPU times: user 151 ms, sys: 3.77 s, total: 3.92 s
Wall time: 1min 21s
```

Comment (cpcloud) [inline odo/backends/sparksql.py]: This is on a file of about 3GB of JSON

**Comment by cpcloud** [comment]
@mrocklin addressed your comments

**Comment by cpcloud** [comment]
firing on all cylinders!

**Comment by cpcloud** [comment]
@mrocklin any more comments?

**Comment by mrocklin** [comment]
+1

**Pull request merged**